Winrm commands stdin, stdout, stderr; Commands Live tests

bostko opened the pull request:
This PR provides stdin, stderr and stdout support to the Windows commands.
Look the commits to see the changes

**Not** ready for review.

It has a Live tests for it and there is an example blueprint for windows byon location which I can remove from the PR if you'd like.

I also changed the config key ByonLocationResolver.OS_FAMILY. Here are my comments on these 
So JcloudsLocationConfig.OS_FAMILY config key is osFamily and for ByonLocationResolver.OS_FAMILY it is osfamily without capital F
I was thinking for fixing it by moving this config key to AbstractLocation. Is it possible that we don't want to include this config to all descendants of AbstractLocation?

aledsage commented on core/src/main/java/org/apache/brooklyn/location/byon/ByonLocationResolver.java:
```diff
@@ -199,8 +199,8 @@ protected ConfigBag extractConfig(Map<?,?> locationFlags, String spec, LocationR
-        if (osfamily != null) {
-            locationClassHere = getLocationClass(osfamily);
+        if (osFamily != null) {
+            locationClassHere = OS_TO_MACHINE_LOCATION_TYPE.get(osFamily);
```
This reverts @sjcorbett 's change in commit c3e027b80826427d04e3b726bf9f71f24fc46103. Can you change it back to use `getLocationClass(osFamily)` please.

aledsage commented on core/src/test/java/org/apache/brooklyn/location/byon/ByonLocationResolverTest.java:
```diff
@@ -345,7 +345,7 @@ public void testWindowsMachines(String osFamily) throws Exception {
         Map<String, ?> flags = ImmutableMap.of(
                 "hosts", ImmutableList.of("1.1.1.1", "2.2.2.2"),
-                "osfamily", osFamily
+                "osFamily", "windows"
```
This reverts Sam's change (where he tested it is case insensitive). Can you revert to use the `osFamily` parameter please.

bostko replied:
I am doing exactly this.
I've changed it from osfamily to osFamily.
Notice that this was only in ByonLocationResolver. Can you check my PR comment on top. What do you think for moving this parameter in AbstractLocation

aledsage replied:
Please use the `osFamily` parameter that is being passed in - i.e. `"osFamily", osFamily`

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessDriver.java:
```diff
@@ -116,11 +116,9 @@ public void start() {
-                    runPreInstallCommand(entity.getConfig(BrooklynConfigKeys.PRE_INSTALL_COMMAND));
-                }});
-            };
+            DynamicTasks.queue("pre-install-command", new Runnable() { public void run() {
```
Seems wrong to have "pre-install-command" as a special case. There is still a check that the command is non-blank for `POST_INSTALL_COMMAND`, `PRE_LAUNCH_COMMAND` and `POST_LAUNCH_COMMAND`. We should either include the if check for all of them or for none of them.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessDriver.java:
```diff
@@ -189,7 +187,7 @@ public void start() {
     public void preInstall() {}
 
-    public abstract void runPreInstallCommand(String command);
+    public abstract void runPreInstallCommand();
```
Same here: the `runPreInstallCommand` should have the same signature as `runPostInstallCommand` etc.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessWinRmDriver.java:
```diff
@@ -39,8 +44,11 @@
+import org.slf4j.LoggerFactory;
 
 public abstract class AbstractSoftwareProcessWinRmDriver extends AbstractSoftwareProcessDriver {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractSoftwareProcessDriver.class);
```
logger should say `AbstractSoftwareProcessWinRmDriver.class`

aledsage replied:
Please change the logger to `LoggerFactory.getLogger(AbstractSoftwareProcessWinRmDriver.class)`

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessWinRmDriver.java:
```diff
@@ -129,10 +132,40 @@ protected WinRmToolResponse executeCommand(ConfigKey<String> regularCommandKey,
+            response = getLocation().executeScript(ImmutableList.of(regularCommand));
+        }
+
+        if (Tasks.current()!=null) {
```
I'd be inclined to add the `tagDynamicallyWithSoftStream` before executing the script. Then after the script has executed, update that stream with the contents of `response.getStdOut()` etc.

If you did that, then while the command is executing you'd be able to inspect the stdin of the command and you'd be able to see that there will be a stdout/stderr (but they would be empty). The alternative here would be surprising for users (i.e. in the activity view, it does not show that there will be any streams; then when you navigate away from the task and back again once it's finished it shows that there are now streams).

aledsage replied:
I'd cache whether we've made use of these stdOut/stdErr, rather than calling `if (Tasks.current()!=null)` again. If we added those tags previously then call `writeToStream` (rather than also looking up the `BrooklynTaskTags.stream()` again - no need to do that).

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessWinRmDriver.java:
```diff
@@ -129,10 +132,40 @@ protected WinRmToolResponse executeCommand(ConfigKey<String> regularCommandKey,
+
+            Tasks.addTagDynamically(BrooklynTaskTags.tagForStreamSoft(brooklynTaskTag, stdStream));
+
+            if(LOG.isDebugEnabled()) LOG.debug("{} {}", brooklynTaskTag, std);
```
To double-check: is this the only place the stdout / stderr is being logged? Or will we now have multiple copies of stdout being logged? We only want one copy in the log.

It's also definitely worth including more context in the log message. If it just says "stdout ...." then you can't tell which entity it was for or which machine it was on. Therefore include a prefix. For example, here is a log statement for ssh stdout (which is written one line at a time):

    2015-08-18 12:58:13,400 DEBUG brooklyn.SSH [Thread-50601]: [amp@<ip>:22:stdout] WARNING: no known/successful package manager to install iptables-persistent, may fail subsequently

bostko commented:
I will appreciate if anyone review it earlier.
Thank you!

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessWinRmDriver.java:
```diff
@@ -129,10 +132,40 @@ protected WinRmToolResponse executeCommand(ConfigKey<String> regularCommandKey,
+
+            if(LOG.isDebugEnabled()) LOG.debug("{} {}", brooklynTaskTag, std);
+        } catch (IOException e) {
+            LOG.warn(Joiner.on('\n').join(e.getStackTrace()));
```
I'd log with `LOG.warn("Problem populating "+brooklynTaskTag+" for task of entity "+getEntity(), e);`

It's better to let the logging framework know it's an exception, and to give more context for what was being done when the exception was thrown.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/VanillaWindowsProcessWinRmDriver.java:
```diff
@@ -18,12 +18,17 @@
+import org.slf4j.LoggerFactory;
 
 public class VanillaWindowsProcessWinRmDriver extends AbstractSoftwareProcessWinRmDriver implements VanillaWindowsProcessDriver {
+    Logger LOG = LoggerFactory.getLogger(VanillaWindowsProcessWinRmDriver.class);
```
Always declare these `private static final Logger LOG =`.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/VanillaWindowsProcessWinRmDriver.java:
```diff
@@ -37,31 +42,36 @@ public void start() {
+    public void runPreInstallCommand() {
+        if(Strings.isNonBlank(getEntity().getConfig(VanillaWindowsProcess.PRE_INSTALL_COMMAND)) || Strings.isNonBlank(getEntity().getConfig(VanillaWindowsProcess.PRE_INSTALL_POWERSHELL_COMMAND))) {
+            executeCommand(VanillaWindowsProcess.PRE_INSTALL_COMMAND, VanillaWindowsProcess.PRE_INSTALL_POWERSHELL_COMMAND, true);
+            if (entity.getConfig(VanillaWindowsProcess.PRE_INSTALL_REBOOT_REQUIRED)) {
```
If one had configured `PRE_INSTALL_REBOOT_REQUIRED` without any pre install command, we should still execute it. Let's move this check outside of the outer if block.

Also, I'd use `if (Boolean.TRUE.equals(entity.getConfig(VanillaWindowsProcess.PRE_INSTALL_REBOOT_REQUIRED))) {`. Otherwise there is a risk of a NullPointerException if the config parameter's value is null.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/VanillaWindowsProcessWinRmDriver.java:
```diff
@@ -37,31 +42,36 @@ public void start() {
-            rebootAndWait();
+        if(Strings.isNonBlank(getEntity().getConfig(VanillaWindowsProcess.INSTALL_COMMAND)) || Strings.isNonBlank(getEntity().getConfig(VanillaWindowsProcess.INSTALL_POWERSHELL_COMMAND))) {
+            executeCommand(VanillaWindowsProcess.INSTALL_COMMAND, VanillaWindowsProcess.INSTALL_POWERSHELL_COMMAND, true);
+            if (entity.getConfig(VanillaWindowsProcess.INSTALL_REBOOT_REQUIRED)) {
```
Same here and in the other commands: move the reboot check outside of the outer if block.

aledsage commented on software/base/src/test/java/org/apache/brooklyn/entity/software/base/VanillaSoftwareProcessStreamsIntegrationTest.java:
```diff
@@ -59,8 +39,24 @@ public void setUp() throws Exception {
     @Test(groups = "Integration")
+    @Override
     public void testGetsStreams() {
-        Map<String, String> cmds = ImmutableMap.<String, String>builder()
```
Slightly simpler to do `Map<String, String> cmds = getCommands();`. And then below to use `cmds` rather than repeatedly calling `getCommands()`.

aledsage commented on software/base/src/test/java/org/apache/brooklyn/entity/software/base/VanillaSoftwareProcessStreamsIntegrationTest.java:
```diff
@@ -59,8 +39,24 @@ public void setUp() throws Exception {
+                .configure(VanillaSoftwareProcess.POST_LAUNCH_COMMAND, "echo " + getCommands().get("post-launch-command"))
+                .configure(VanillaSoftwareProcess.CHECK_RUNNING_COMMAND, "echo true"));
+        app.start(ImmutableList.of(localhost));
+        assertStreams(entity);
```
I'd pass the `cmds` as an argument of `assertStreams(...)`. That slightly decreases the coupling between the various methods. (very minor point).

aledsage commented on software/base/src/test/java/org/apache/brooklyn/entity/software/base/test/location/WinRmMachineLocationLiveTest.java:
```diff
@@ -65,7 +65,7 @@ public void setUpClass() throws Exception {
                 "inboundPorts", ImmutableList.of(5985, 3389),
                 "displayName", "AWS Oregon (Windows)",
-                "imageId", "us-west-2/ami-8fd3f9bf",
+                "imageNameRegex", ".*Windows.*2012.*",
```
A customer said that they also needed to specify an image owner, in order to avoid a marketplace windows AMI being picked. Can you search for AMIs that match that regex to see if that is a risk?

For example:

    brooklyn.location.named.mywindows1.imageOwner=801119661308
    brooklyn.location.named.mywindows1.imageNameRegex=Windows_Server-2012-R2_RTM-English-64Bit-Base-.*

This is used within our jclouds integration code (in ComputeServiceRegistryImpl.findComputeService) where we have:

   properties.setProperty(PROPERTY_EC2_AMI_QUERY,
   "owner-id="+conf.getStringKey("imageOwner")+";state=available;image-type=machine");

(It's a bit ugly having the code there; would be nicer if that were a first-class property on jclouds' AWSEC2TemplateOptions. That's an improvement we could make in jclouds. It would also be nice to have a config key for `imageOwner` rather than just this hard-coded constant being looked up in the config!).

aledsage commented on software/base/src/test/resources/brooklyn/entity/basic/fast_stdout_blueprint_test.yaml:
```diff
@@ -0,0 +1,40 @@
+# specific language governing permissions and limitations
+# under the License.
+#
+name: Windows stdout
```
Not sure about having this yaml file here, given it is never used.

If you do leave it in, then it definitely needs a good comment explaining what it does and what it is useful for.

aledsage commented:
Thanks @bostko - finished reviewing.

bostko commented:
@aledsage Thank you for the elaborative comments!
I fixed the addressed issues and added new config for powershell commands for: post-install, pre-launch, post-launch
You can review it again now.

aledsage commented:
Build failure by asfbot was for `RebindJmxFeedTest.testJmxFeedIsPersisted` with "Port already in use: 40131", so is unrelated to this PR.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/VanillaWindowsProcessWinRmDriver.java:
```diff
@@ -72,8 +72,12 @@ public void launch() {
+        WinRmToolResponse runningCheck = executeCommand(VanillaWindowsProcess.CHECK_RUNNING_COMMAND,
+                VanillaWindowsProcess.CHECK_RUNNING_POWERSHELL_COMMAND, false);
+        if(runningCheck.getStatusCode() != 0) {
+            LOG.info(this + " isRunning check failed: exit code "  + runningCheck.getStatusCode() + "; " + runningCheck.getStdErr());
```
Instead of logging `this`, I'd log the `getEntity()` (the toString on the driver might not be good).

aledsage commented:
@bostko looks good - a few more minor comments.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessWinRmDriver.java:
```diff
@@ -119,20 +136,53 @@ protected void createDirectory(String directoryName, String summaryForLogging) {
+        }
+
+        if (tagWithStd) {
+            if (BrooklynTaskTags.stream(Tasks.current(), BrooklynTaskTags.STREAM_STDOUT)==null) {
```
I don't think you want this additional `if`. I'm surprised it works with it. We've added a stream to this task previously, so when you check if the current task has a null stdout stream, I'd expect it to return false.

Also, you're just writing to your stdout / stderr streams. If someone else has replaced the tag for some reason, then you writing to your streams won't affect that.

bostko replied:
I've followed the pattern from here https://github.com/apache/incubator-brooklyn/blob/master/software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessSshDriver.java#L262
There it is used check for STDOUT availability

aledsage replied:
But the way you've followed that is different. In https://github.com/apache/incubator-brooklyn/blob/master/software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessSshDriver.java#L262 it has the `if` guard for the call to `Tasks.addTagDynamically`. You have done that a few lines above.

My point here is that once you've called `Tasks.addTagDyamically` a few lines above, you don't want to then do the `if` guard again for when you call `writeToStream` - I'd expect that second if to return false because you previously added the steam tag for this task.

aledsage commented:
I'll merge this now, and then change the line in `AbstractSoftwareProcessWinRmDriver` as discussed in the comments.
